### PR TITLE
Use defer to cleanup allocated service ip in case of error

### PR DESCRIPTION
**Pull request by justinsb**
We're probably going to be doing a port allocation here also (#5225), so this needs some pre-simplifying.

Also interested to hear if this is the best approach.

### CODE REVIEW

File: pkg/registry/service/rest.go [hunk 1]
```diff
@@ -97,6 +97,15 @@ func (rs *REST) Create(ctx api.Context, obj runtime.Object) (runtime.Object, err
 		return nil, err
 	}
 
+	releaseServiceIp := false
```
Comment (lavalamp) [inline pkg/registry/service/rest.go]: nit: releaseServiceIP

**Comment by lavalamp** [comment]
LGTM, one nit

**Comment by thockin** [comment]
I'm fine with this.  I have in the past used a cancellable function-caller, but it's not as clean without C++-style destruction.

I have not used Go lambda-capture much, but I find this less obvious than it might be.  For consideration and shootdown:

```
defer func(release *bool) { /* code */}(&releaseIP)
```

Better or worse?

**Comment by satnam6502** [comment]
In Justin's code `releaseServiceIp` will be captured by reference and I think the right thing happens. I think what Tim proposes achieves the same thing by passing in a reference to `releaseIP` but I think Justin's original code is fine. I think arguments to lambda expressions make sense when you want to capture a value rather than a reference.

**Comment by justinsb** [comment]
One of the reasons for putting this through as a separate (small) PR was that I thought there might be style discussions!  I don't love the pattern as I wrote it, but I think if we pick a pattern it will be recognizable much like `if err = f(); err != nil {`

For completeness: an alternative is not to declare a separate 'cleanup' variable, but to reuse err  (i.e. we do cleanup `if err != nil`); however I think this will cause bugs because of Go's variable shadowing behaviour.

I can't even remember to type ID instead of Id, so whatever you want is fine by me ;-)

**Comment by thockin** [comment]
I don't think shadowing a common variable would be good.  As it stands it is pretty obvious, even if it is a bit surprising to those of us not deeply familiar with variable capture in Go.  My proposal was to make it EVEN MORE obvious, but maybe so much that it's anti-idiomatic.

File: pkg/registry/service/rest.go (hunk 1 above)
Comment (thockin) [inline pkg/registry/service/rest.go]: +1

**Comment by thockin** [comment]
Fix the naming nit and I'll merge it :)  Thanks.

**Comment by satnam6502** [comment]
I think Tim does have a point here. Personally, I think the way variables are captured in Go is not very different from what happens in other imperative languages but it can catch you out if you're not paying attention. Tim's proposal does make things crystal clear for the casual observer.

**Comment by justinsb** [comment]
I've fixed ServiceIp -> ServiceIP.  Also put up a WIP of the work that this was part of (cluster-wide assignment of ports) at #6684

**Comment by thockin** [comment]
LGTM

**Comment by justinsb** [comment]
BTW, I absolutely agree with Tim as well.. but I think this is the most godiomatic way.  Until we're ready to start our own idioms.  (And when we are, I'd like to propose that "id" be a word, not an acronym, as in http://en.wikipedia.org/wiki/Id,_ego_and_super-ego )

**Pull request merged**